### DISCUSSION TITLE
How to implement this pre-push git hook in nushell ?

### DISCUSSION THREAD

**Discussion by hustcer** [nushell/nushell#4852] [Q&A]
I'm using [husky](https://github.com/typicode/husky) to manage a `pre-push` git hook, It's implemented in `sh` and then invoke a `nushell` script as bellow:
```sh
#!/bin/sh
# Git pre push hooks, need just/nu and related nu scripts to run.

# set -exo pipefail

. "$(dirname "$0")/_/husky.sh"

while read local_ref local_oid remote_ref remote_oid
do

  if ! command -v just &> /dev/null; then
    echo "Command 'just' could not be found, Please install it by 'brew install just', and try again!\n"
    break;
  fi
  if ! command -v nu &> /dev/null; then
    echo "Command 'nu' could not be found, Please install it by 'brew install nushell', and try again!\n"
    break;
  fi

  if [[ $local_ref == '(delete)' ]]; then local_ref='_delete_'; fi
  just --justfile ~/.justfile --dotenv-path ~/.env git-sync-branch $local_ref $local_oid $remote_ref;
  # Break is important here, to stop another loop
  break;
done

exit 0

```
`git-sync-branch` command was written by nushell, and I'm wondering can I implement this `pre-push` hook in nushell? It seems that we don't have sth. like `while read ...` and `break`?

**Comment by kubouch** [comment]
We don't have a lot of control flow options right now but you can simulate `break` using `reduce -f`, something like this maybe: 
```
[ local_ref local_oid remote_ref remote_oid ] | reduce -f true {|it, acc| 
    if $acc {
        ... run code        
        false # pass $false to future iterations of $acc which will skip the loop iteration
    } else {
        false # pass $false to future iterations of $acc which will skip the loop iteration
    }
}
```

Sourcing from `husky.sh` also won't work. We were thinking about capturing environment of an existing process which would allow this but haven't been successful implementing it yet.

**Reply by hustcer** [reply]
Thanks, Then how about `while read`?

**Reply by kubouch** [reply]
I'm not sure what `while read` is supposed to do. How is it different from a simple iteration?

**Reply by hustcer** [reply]
> I'm not sure what `while read` is supposed to do. How is it different from a simple iteration?

Maybe it's used for watching all the following push, so that users can do another push before the previous push hook task finished, my guess.
